test/messenger.test.ts

```diff
@@ -0,0 +1,177 @@
+import Messenger from '../src/messenger';
+import { EmitterMessage, WorkerMessage } from '../src/types';
+
+describe('Messenger', () => {
+  describe('constructor', () => {
+    it('creates and connects to a shared worker', () => {
+      const sharedWorkerUrl = 'http://foo.bar.org';
+      const mockStart = jest.fn();
+      SharedWorker = jest.fn(
+        () =>
+          ({
+            port: {
+              start: mockStart,
+              onmessage: undefined,
+            },
+          } as any)
+      );
+      const handleWorkerSpy = jest.spyOn(
+        Messenger.prototype,
+        'handleWorkerMessage'
+      );
+      handleWorkerSpy.mockImplementation(jest.fn());
+      const messenger = new Messenger({
+        sharedWorkerUrl,
+      });
+      expect(SharedWorker).toHaveBeenCalledWith(sharedWorkerUrl);
+      expect(mockStart).toHaveBeenCalled();
+      (messenger['worker'] as any)['port']['onmessage']({ data: 'foo' });
+      expect(handleWorkerSpy).toHaveBeenCalled();
+      handleWorkerSpy.mockRestore();
+    });
+
+    it.each([
+      [true, '/worker.js', true],
+      [false, undefined, true],
+      [false, '/worker.js', false],
+    ])(
+      'does not create a shared worker when it is disabled, no url is passed, or unavailable',
+      (_disableSharedWorker, sharedWorkerUrl, sharedWorkerAvailable) => {
+        SharedWorker = sharedWorkerAvailable ? jest.fn() : (undefined as any);
+        const messenger = new Messenger({
+          sharedWorkerUrl,
+          _disableSharedWorker,
+        });
+        expect(messenger['worker']).toBeUndefined();
+      }
+    );
+  });
+
+  describe('message emitter', () => {
+    let messenger: Messenger;
+    const name1 = 'connect';
+    const name2 = 'disconnect';
+    const fn1 = (a: any) => {};
+    const fn2 = (b: any) => {};
+
+    beforeEach(() => {
+      messenger = new Messenger({});
+    });
+
+    it('can add listeners', () => {
```

Comment:
These tests seem to be testing the internal implementation heavily. It'll be better to test the actual triggers and then once removed, the functions are no longer triggered.

Reply:
haha yeah fair enough 😄

Reply:
i think i'm gonna skip doing the same with the `SharedWorker` though, since it's not defined in the test environment and we'd have to stub implementation details